sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala

```diff
@@ -0,0 +1,202 @@
+ *                   to the relativeSD used when writing the HLL++ sketch.
+ */
+// scalastyle:on
+@ExpressionDescription(
+  usage = """
+    _FUNC_(expr) - Returns the estimated cardinality of the HyperLogLog++ sketch.""",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(approx_count_distinct_sketch(col1))
+        FROM VALUES (1), (1), (2), (2), (3) tab(col1);
+       3
+  """,
+  group = "agg_funcs",
+  since = "3.3.1")
+case class HyperLogLogPlusPlusAggSketch(
+    child: Expression,
+    var relativeSD: Double = HyperLogLogPlusPlus.defaultRelativeSD,
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0) extends HyperLogLogPlusPlusTrait {
+
+  def this(child: Expression) = {
+    this(
+      child = child,
+      relativeSD = HyperLogLogPlusPlus.defaultRelativeSD,
+      mutableAggBufferOffset = 0,
+      inputAggBufferOffset = 0
+    )
+  }
+
+  def this(child: Expression, relativeSD: Expression) = {
+    this(
+      child = child,
+      relativeSD = HyperLogLogPlusPlus.validateDoubleLiteral(relativeSD),
+      mutableAggBufferOffset = 0,
+      inputAggBufferOffset = 0
+    )
+  }
+
+  override def prettyName: String = "approx_count_distinct_agg_sketch"
+
+  /**
+   * Allocate enough words to store all registers, plus one extra long
+   * allowing us to check buffer compatibility during merge. This is mildly
+   * different from the HyperLogLogPlusPlus trait which doesn't store numWords
+   * at the head of the aggregation buffer.
+   */
+  override val aggBufferAttributes: Seq[AttributeReference] = {
+    Seq.tabulate(hllppHelper.numWords + 1) { i =>
+      AttributeReference(s"MS[$i]", LongType)()
+    }
+  }
+
+  // Note: although this simply copies aggBufferAttributes, this common code can not be placed
+  // in the superclass because that will lead to initialization ordering issues.
+  override val inputAggBufferAttributes: Seq[AttributeReference] =
+  aggBufferAttributes.map(_.newInstance())
+
+  /**
+   * This aggregate merges the HLL++ sketches that were written by a previous
+   * HyperLogLogPlusPlusSketch instance, which utilized a HLL++ helper that may have been
+   * configured differently than the one provided by our super class. We'll maintain the
+   * relativeSD and create a new helper for the first sketch that we deserialize, ensuring
+   * that every following sketch is aggregateable and can fit within the original agg buffer.
+   */
+  private var relativeSDUpdated: Boolean = false
+
+  /**
+   * Update the HLL++ buffer; input is expected to be the serialized
+   * HLL++ sketch, which we'll merge into the shared agg buffer.
+   */
+  override def update(buffer: InternalRow, input: InternalRow): Unit = {
+    val v = child.eval(input)
+    if (v != null) {
+
+      val (newRelativeSD, newInternalRow) =
+        HyperLogLogPlusPlusSketch.deserializeSketch(v.asInstanceOf[Array[Byte]])
+
+      relativeSDUpdated match {
+        case true =>
+          if (newRelativeSD != relativeSD) {
```

Comment:
here we are comparing floating-point numbers for exact inequality. Are we certain there are no false positives/negatives for this? Same for the `if (newRelativeSD < relativeSD)` below?

Reply:
I modified this to test the Integer numWords values derived from the Double relativeSD values so we don't have to implement some custom precision comparison functions.